packages/salesforcedx-vscode-soql/src/queryDataView/queryDataFileService.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { JsonMap } from '@salesforce/ts-types';
+import * as fs from 'fs';
+import { QueryResult } from 'jsforce';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { QUERY_RESULTS_DIR_PATH } from '../constants';
+import {
+  CsvDataProvider,
+  DataProvider,
+  JsonDataProvider
+} from './dataProviders';
+
+export enum FileFormat {
+  JSON = 'json',
+  CSV = 'csv'
+}
+
+const sfdxCoreExtension = vscode.extensions.getExtension(
+  'salesforce.salesforcedx-vscode-core'
+);
+const sfdxCoreExports = sfdxCoreExtension
+  ? sfdxCoreExtension.exports
+  : undefined;
+const { getRootWorkspacePath } = sfdxCoreExports;
+
+export class QueryDataFileService {
+  private dataProvider: DataProvider;
+
+  constructor(
+    private queryData: QueryResult<JsonMap>,
+    private format: FileFormat,
+    private documentName: string
+  ) {
+    this.dataProvider = this.getDataProvider();
+  }
+
+  protected getDataProvider(): DataProvider {
+    switch (this.format) {
+      case FileFormat.CSV:
+        return new CsvDataProvider(this.documentName);
+      case FileFormat.JSON:
+        return new JsonDataProvider(this.documentName);
+      default:
+        throw new Error('No DataProvider Found');
+    }
+  }
+
+  public save(): string {
+    try {
+      const fileContent = this.dataProvider.getFileContent(
+        this.queryData.records
+      );
+      const savedFileName = this.dataProvider.getFileName();
+      const queryDataFilePath = path.join(
+        this.getResultsDirectoryPath(),
+        savedFileName
+      );
+
+      this.createResultsDirectoryIfDoesNotExist();
+      // Save query results to disk
+      fs.writeFileSync(queryDataFilePath, fileContent);
+      this.showSaveSuccessMessage(savedFileName);
+      this.showFileInExplorer(queryDataFilePath);
+
+      return queryDataFilePath;
+    } catch (error) {
+      // TODO: i18n, CCX
+      vscode.window.showErrorMessage(
+        `Your data could not be saved. Run the query and try again.`
+      );
+      throw error;
+    }
+  }
+
+  private getResultsDirectoryPath() {
+    return path.join(getRootWorkspacePath(), QUERY_RESULTS_DIR_PATH);
+  }
+
+  private createResultsDirectoryIfDoesNotExist() {
+    fs.mkdirSync(this.getResultsDirectoryPath(), {
+      recursive: true
+    });
+  }
+
+  private showFileInExplorer(targetPath: string) {
+    vscode.commands.executeCommand(
+      'revealInExplorer',
+      vscode.Uri.parse(targetPath)
+    );
+  }
+
+  private showSaveSuccessMessage(savedFileName: string) {
+    vscode.window.showInformationMessage(
+      // TODO: i18n and CCX
+      `Your data has been saved in this workspace as: ${savedFileName}`
```

Comment:
@ekapner Sorry I forgot to mention you here.

This message will be shown to the user when the file is successfully saved in the `query-results` directory.
I was thinking to at least change from "data" to "results" to be consistant.
> `Your results have been saved in this workspace as: ${savedFileName}`

Reply:
@jonnyhork - should we say "query-results" folder instead of "workspace"?

We saved the results in the query-results folder as: ${savedFileName}

Reply:
yeah thats fine with me

Reply:
@jonnyhork - changed my mind. :)
The results were saved in the query-results folder as: $...